### ISSUE TITLE
Roslyn reports inconsistent diagnostics for T and T? values assigned to [NotNull, MaybeNull] return types

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#51509]
**Version Used**: 
```
Branch master (9 Feb 2021)
Latest commit 2d3ffe0 by msftbot[bot]:
Merge pull request #49990 from ryzngard/feature/namespace_analyzer

Add sync namespace analyzer
```

**Steps to Reproduce**:

Compile the following code
```cs
using System.Diagnostics.CodeAnalysis;

class C<T>
{  
  [return: NotNull, MaybeNull] public T M1(T t) => t; // CS8607
  [return: NotNull, MaybeNull] public T M2(T? t) => t; // no warnings
  
  [return: NotNull, MaybeNull] public T M3(T t) { t = default; return t; } // CS8600 for 't = default', no warnings for 'return t'
  [return: NotNull, MaybeNull] public T M4(T? t) { t = default; return t; } // no warnings
}
```

**Expected Behavior**:
Either all methods have warnings for possible nullable value returned from `[return: NotNull]` method or neither of them.

**Actual Behavior**:
`CS8607: A possible null value may not be used for a type marked with [NotNull] or [DisallowNull]` is reported for `M1` but not in other methods even though `M3` and `M4` explicitly assign a nullable state to the returned variable.

**Notes**
Both `T` and `T?` can contain null values. However Roslyn seems to ignore some combination of nullable state and `[NotNull, MaybeNull]` annotations.

**Comment by xtqqczze** [comment]
The issue description does not account for nullability of generics, see [Nullable reference types § Generics](https://learn.microsoft.com/dotnet/csharp/nullable-references#generics).
